### PR TITLE
Dynamically generate SQLDelight definitions

### PR DESCRIPTION
**Pull request by MMarco94**
With this PR, the `.sq` files related to the Tmdb cache will be generated automatically. 
The generated files are kept in Git, as their definition should be stable, and any change needs to be reviewed. 

An alternative to having a table per API, is to batch APIs together based on their "key" (e.g. one table that caches all APIs related to a movie). In practice, the batching wouldn't solve much, as few APIs share the same "key".

In future PRs, we can go a step further and also generate the SQLDelight table adapters.

### CODE REVIEW

File: buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt [hunk 1]
```diff
@@ -0,0 +1,110 @@
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.TaskAction
+import java.io.File
+
+/**
+ * A Gradle Task to generate the SQLDelight definitions related to the Tmdb Cache.
+ */
+open class SqlDelightTmdbCacheDefinitionsGenerator : DefaultTask() {
+    /** Where to generate the .sq files */
+    private val sqldelightRoot = File(project.projectDir, "src/androidMain/sqldelight/tmdbCache/io/github/couchtracker/db/tmdbCache")
```
Comment (MMauro94) [inline buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt:10]: Can we use the Path API instead? 

You can also do fun stuff like:
```kotlin
private val sqldelightRoot = project.projectDir.toPath() / "src" / "androidMain" / "sqldelight" / "tmdbCache" / "io" / "github" / "couchtracker" / "db" / "tmdbCache"
```

File: buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    private data class SqlFunction(
+        val name: String,
+        val content: StringBuilder.(SqlTable) -> Unit,
+    )
+
+    @TaskAction
+    fun run() {
+        caches.forEach { cache ->
```
Comment (MMauro94) [inline buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt:96]: Per Kotin-style guidelines, regular for should be preferred over forEach: https://kotlinlang.org/docs/coding-conventions.html#loops

Same thing for other parts of the code.

File: buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+/**
+ * A Gradle Task to generate the SQLDelight definitions related to the Tmdb Cache.
+ */
+open class SqlDelightTmdbCacheDefinitionsGenerator : DefaultTask() {
+    /** Where to generate the .sq files */
+    private val sqldelightRoot = File(project.projectDir, "src/androidMain/sqldelight/tmdbCache/io/github/couchtracker/db/tmdbCache")
+
+    /** Common columns */
+    private val movieIdColumn = SqlColumn.int("tmdbId", "io.github.couchtracker.tmdb.TmdbMovieId")
+    private val languageColumn = SqlColumn.text("language", "io.github.couchtracker.tmdb.TmdbLanguage")
```
Comment (MMauro94) [inline buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt:14]: These should be constants

File: buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            name = "MovieReleaseDatesCache",
+            key = listOf(movieIdColumn),
+            value = SqlColumn.text("releaseDates", "kotlin.collections.List<app.moviebase.tmdb.model.TmdbReleaseDates>"),
+        ),
+    )
+
+    /** Which sql functions to generate */
+    private val sqlFunctions = listOf(
+        SqlFunction("get") { cache ->
+            append("SELECT ${cache.value.name}\n")
```
Comment (MMauro94) [inline buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt:33]: Can we use [appendLine](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/append-line.html) here? It avoids having to remember the ugly \n at the end of each line

File: buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+    init {
+        group = "io.github.couchtracker"
+        description = "Generates the SqlDelight definitions for local the Tmdb Cache"
+    }
+
+    private data class SqlColumn(
+        val name: String,
+        val sqlType: String,
+        val kotlinType: String,
```
Comment (MMauro94) [inline buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt:52]: Can we use a KType here instead? It retains generic type info: https://pl.kotl.in/iDHs2HQpC (we may need to add a dependncy to kotlin-reflect). This allows to have more complile-time safeties.

It will also avoid the need for the nullable boolean.

**Comment by MMauro94** [comment]
>In future PRs, we can go a step further and also generate the SQLDelight table adapters.

We can use KotlinPoet for that: https://square.github.io/kotlinpoet/

File: composeApp/src/androidMain/kotlin/io/github/couchtracker/tmdb/TmdbMovie.kt [hunk 6]
```diff
@@ -22,12 +22,12 @@ data class TmdbMovie(
...
     suspend fun releaseDates(cache: TmdbCache): List<TmdbReleaseDates> = tmdbGetOrDownload(
-        get = { cache.movieReleaseDatesCacheQueries.get(id, language).awaitAsOneOrNull() },
-        put = { cache.movieReleaseDatesCacheQueries.put(tmdbId = id, language = language, releaseDates = it) },
+        get = { cache.movieReleaseDatesCacheQueries.get(id).awaitAsOneOrNull() },
+        put = { cache.movieReleaseDatesCacheQueries.put(tmdbId = id, releaseDates = it) },
         downloader = {
             it.movies.getDetails(
                 id.value,
-                language.apiParameter,
+                null,
```
Comment (MMauro94) [inline composeApp/src/androidMain/kotlin/io/github/couchtracker/tmdb/TmdbMovie.kt:30]: Why this change? Does the language not change the output of this API?

File: buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt (hunk 1 above)
Comment (MMarco94) [reply buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt:10]: Why?

File: buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt (hunk 3 above)
Comment (MMarco94) [reply buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt:14]: Detekt says:
> detekt - VariableNaming: Private variable names should match the pattern: (_)?[a-z][A-Za-z0-9]*

File: buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt (hunk 5 above)
Comment (MMarco94) [reply buildSrc/src/main/kotlin/SqlDelightTmdbCacheDefinitionsGenerator.kt:52]: No, this module cannot depend on the main module, so we cannot have static references to the classes

File: composeApp/src/androidMain/kotlin/io/github/couchtracker/tmdb/TmdbMovie.kt (hunk 6 above)
Comment (MMarco94) [reply composeApp/src/androidMain/kotlin/io/github/couchtracker/tmdb/TmdbMovie.kt:30]: Exactly. We might need to rethink this `TmdbMovie`, since it doesn't make much sense to have this here, but I'm going to leave this to another PR

**Review by MMauro94 — APPROVED** [review]
